gk/main.c

```diff
@@ -684,92 +684,16 @@ setup_gk_instance(unsigned int lcore_id, struct gk_config *gk_conf)
 	return ret;
 }
 
-static struct flow_entry *
-find_flow_entry_candidate(struct gk_instance *instance,
-	uint32_t bidx, enum gk_flow_state state_to_add)
-{
-	int32_t index;
-	uint32_t next = 0;
-	const struct ip_flow *key;
-	struct flow_entry *last_fe = NULL;
-	void *data;
-	uint64_t now = rte_rdtsc();
-
-	index = rte_hash_bucket_iterate(instance->ip_flow_hash_table,
-		(void *)&key, &data, &bidx, &next);
-	while (index >= 0) {
-		struct flow_entry *fe = &instance->ip_flow_entry_table[index];
-
-		/* Expired flow entry. */
-		if (is_flow_expired(fe, now))
-			return fe;
-
-		/*
-		 * Only flow entries with state GK_REQUEST will possibly
-		 * be repaced; others have a higher priority.
-		 */
-		if (fe->state == GK_REQUEST &&
-				state_to_add != GK_REQUEST &&
-				(last_fe == NULL ||
-				last_fe->u.request.last_packet_seen_at >
-				fe->u.request.last_packet_seen_at))
-			last_fe = fe;
-
-		index = rte_hash_bucket_iterate(instance->ip_flow_hash_table,
-			(void *)&key, &data, &bidx, &next);
-	}
-
-	return last_fe;
-}
-
-static int
-drop_flow_entry_heuristically(struct gk_instance *instance,
-	hash_sig_t sig, enum gk_flow_state state_to_add)
-{
-	uint32_t primary_bidx = rte_hash_get_primary_bucket(
-		instance->ip_flow_hash_table, sig);
-	struct flow_entry *fe = find_flow_entry_candidate(
-		instance, primary_bidx, state_to_add);
-	if (fe == NULL)
-		return -ENOSPC;
-
-	return gk_del_flow_entry_from_hash(instance->ip_flow_hash_table, fe);
-}
-
-/*
- * We heuristically drop entries to alleviate memory pressure
- * when the table is full.
- */
 static int
 gk_hash_add_flow_entry(struct gk_instance *instance,
-	struct ip_flow *flow, uint32_t rss_hash_val,
-	enum gk_flow_state state_to_add)
+	struct ip_flow *flow, uint32_t rss_hash_val)
 {
-	int retried = false;
-	while (true) {
-		int ret = rte_hash_add_key_with_hash(
-			instance->ip_flow_hash_table, flow, rss_hash_val);
-		if (ret == -ENOSPC) {
-			RTE_VERIFY(!retried);
-			if (state_to_add == GK_REQUEST)
-				return ret;
-
-			ret = drop_flow_entry_heuristically(instance,
-				rss_hash_val, state_to_add);
-			if (ret < 0)
-				return -ENOSPC;
-			retried = true;
-			continue;
-		}
+	int ret = rte_hash_add_key_with_hash(
```

Comment:
Before calling `rte_hash_add_key_with_hash()`, test `if (instance->has_insertion_failed)` and, if true, `return -ENOSPC;`. This will simplify the code of callers.

Notice that this change will create a new performance improvement because it'll apply the skip of new entries to the new policy decisions as well. Therefore, we must run another evaluation.